core/validation/specific_validations/towns_fund_round_four.py

```diff
@@ -296,6 +300,89 @@ def validate_leading_factor_of_delay(
         ]
 
 
+def validate_funding_spent(workbook: dict[str, pd.DataFrame]) -> list["TownsFundRoundFourValidationFailure"] | None:
+    """Validates that total reported funding spent on the sheet is less than the amount specified as allocated in a
+    separate sheet stored on S3.
+
+    For Town deal funds this is done on a per-project basis whereas for future High Street Funds this is done
+    per-programme leading to slightly different logics for both.
+
+    :param workbook: A dictionary where keys are sheet names and values are pandas
+                     DataFrames representing each sheet in the Round 4 submission.
+    :return: ValidationErrors
+    """
+    # pull programme and project indexes from the workbook
+    programme_id = workbook["Programme_Ref"]["Programme ID"][0]
+    fund_type = workbook["Programme_Ref"]["FundType_ID"][0]
+    project_ids = workbook["Project Details"]["Project ID"]
+    funding_df = workbook["Funding"]
+
+    # pull funding spent for individual projects and programme wide total
+    funding_spent = {project: spend_per_project(funding_df, project) for project in project_ids}
+    funding_spent[programme_id] = sum(funding_spent.values())
+
+    funding_allocated = get_allocated_funding()
+
+    funding_spent_failures = []
+    if fund_type == "HS":
+        # check funding against programme wide funding allocated for Future High Street Fund submissions
+        validate_by = "programme"
+        ids_to_check = [programme_id]
+
+    else:
+        # check funding against individual project funding allocated for Towns Deal submissions
+        validate_by = "project"
+        ids_to_check = project_ids.tolist()
+
+    for idx in ids_to_check:
+        if funding_spent[idx] > funding_allocated[idx]:
+            funding_spent_failures.append(
+                TownsFundRoundFourValidationFailure(
+                    sheet="Funding Profiles",
+                    section="Project Funding Profiles"
+                    + ("" if fund_type == "HS" else f" - Project {get_project_number(idx)}"),
```

Comment:
Niggly thing but I think it's more readable if we use `validate_by` here instead of HS? Think it's more intuitive to see the project number being added to the string if `validate_by` == "project" - not immediately clear why fund_type would decide that

Reply:
agree, that helps with readability a lot. I've added this now